src/libraries/Common/src/System/Threading/AsyncOverSyncWithIoCancellation.cs

```diff
@@ -0,0 +1,291 @@
+                // this looping synchronously, we instead queue the invocation of the looping so that it
+                // runs asynchronously from the Cancel call.  Then in order to be able to track its completion,
+                // we store the Task representing that asynchronous work, such that cleanup can wait for the Task.
+                ((AsyncOverSyncWithIoCancellation)s!).CallbackCompleted = Task.Factory.StartNew(static s =>
+                {
+                    var state = (AsyncOverSyncWithIoCancellation)s!;
+
+                    // If cancellation was already requested when UnsafeRegister was called, it'll invoke
+                    // the callback immediately.  If we allowed that to loop until cancellation was successful,
+                    // we'd deadlock, as we'd never perform the very I/O it was waiting for.  As such, if
+                    // the callback is invoked prior to be ready for it, we ignore the callback.
+                    if (!state.FinishedCancellationRegistration)
+                    {
+                        return;
+                    }
+
+                    // Cancel the I/O.  If the cancellation happens too early and we haven't yet initiated
+                    // the synchronous operation, CancelSynchronousIo will fail with ERROR_NOT_FOUND, and
+                    // we'll loop to try again.
+                    SpinWait sw = default;
+                    while (state.ContinueTryingToCancel)
+                    {
+                        if (Interop.Kernel32.CancelSynchronousIo(state.ThreadHandle!))
+                        {
+                            // Successfully canceled I/O.
+                            break;
+                        }
+
+                        if (Marshal.GetLastPInvokeError() != Interop.Errors.ERROR_NOT_FOUND)
+                        {
+                            // Failed to cancel even though there may have been I/O to cancel.
+                            // Attempting to keep trying could result in an infinite loop, so
+                            // give up on trying to cancel.
+                            break;
+                        }
+
+                        sw.SpinOnce();
+                    }
+                }, s, CancellationToken.None, TaskCreationOptions.DenyChildAttach, TaskScheduler.Default);
+            }, this);
+
+            // Now that we've registered with the token, tell the callback it's safe to enter
+            // its cancellation loop if the callback is invoked.
+            FinishedCancellationRegistration = true;
+
+            // And now since cancellation may have been requested and we may have suppressed it
+            // until the previous line, check to see if cancellation has now been requested, and
+            // if it has, stop any callback, remove the registration, and throw.
+            if (cancellationToken.IsCancellationRequested)
+            {
+                reg.Dispose();
```

Comment:
Please excuse me if this is stupid question, but:

the comment says "stop any callback" but the `Dispose` method actually invokes the callback:

https://github.com/dotnet/runtime/blob/96625b4642b33a09d66e13a247e6c377e273065b/src/libraries/Common/src/System/Threading/AsyncOverSyncWithIoCancellation.cs#L172-L174

why is the callback being invoked if we know that we are about to throw?

Reply:
That's not Invoking the callback.